### ISSUE TITLE
calcurse-caldav treat reccuring events of the kind "first tuesday of the month" as "each xth day of the month" where x if the day where the event has been created

### ISSUE THREAD

**Issue by Psycojoker** [lfos/calcurse#207]
Hello,

Step to reproduce:

- on your caldav server that is not calcurse, create an event like "every first tuesday of the month"
- sync using caldav-calcurse
- open calcurse, inside the first even is good, but the next ones aren't on tuesday but on the xth day of the month that was the day of the created even (like they are always on the 7th instead of tuesday)

I've tried looking a bit at the code calcurse-caldav.py but the bug seems located inside calcurse in "recur.c" and it's been like 10 years since I've done C :/

Kind regards,

**Comment by lhca** [comment]
A recurring event of the type you describe cannot be expressed in calcurse at present.

Calcurse is limited to recurring events which can be expressed as "every _n_-th _period_ starting from _day_", where _n_ is a positive integer, _period_ is either day, week, month or year,  and _day_ is the selected day in the calendar. Examples are "every other day", or "every fourth week", or "every month", all starting from a particular day.

Even that is not without its problems and pitfalls. E.g. what does "every month starting from 31 May" mean? Or "every year starting from 28 February"?

**Comment by Psycojoker** [comment]
Thanks for taking the time to write this detailed answer :)

**Comment by lhca** [comment]
Not at all. As an illustration, try creating a recurring appointment like:

`From 22:30 to 22:45 every month starting on 31 May 2019`

In months with 31 days calcurse will show the appointment correctly. But the recurrence rule also generates an appointment for 31 June which by calcurse is interpreted as

`1 July 2019, 22:30 -> 22:45`

which can be recognized as the day after 30 June. But the iCalendar specification (RFC 5545) requires such recurrence instances must be ignored. Calcurse also displays the following instances which are harder to defend:

`1 March 2020, 22:30 -> ..:..`
`2 March 2020, 22:30 -> 22:45`

**Comment by lfos** [comment]
I agree that this is a bug that should be fixed.

If we want to implement the original request, we should first come up with a way to extend the data file format in a clean and backwards compatible way.

**Comment by lhca** [comment]
The instance
 
`2 March 2020, 22:30 -> 22:45`

is calcurse's interpretation of 31 February 2020, i.e. two days after 29 February 2020 (a leap year). But how is

`1 March 2020, 22:30 -> ..:..`

to be explained? Answer: When calcurse looks for recurrent items to show for 1 March 2020, the instance for 31 February is also considered and turned into an instance for 2 March 2020, the day after. But that fact is not recognized, and so an instance for 2 March turns up on the 1 March, explaining the weird end time.

PR #208 makes calcurse ignore impossible dates.

**Comment by lhca** [comment]
> If we want to implement the original request, we should first come up with a way to extend the data file format in a clean and backwards compatible way.

I have studied the recurrence rules in RFC 5545. In their entirety they are a daunting task, but parts of them would be a useful extention to calcurse. I'm thinking of (parts of) BYDAY, BYMONTHDAY and BYMONTH. The basic FREQ, INTERVAL and UNTIL rules are already implemented, and because the others are optional, an extension of the existing data file format that is backwards compatible, should pose no problem.

But there is a major editorial task that should be carrried through before adding new rules. The recurrence variables that describe type, frequency, until date and exception dates, are passed as separate arguments in already very long argument lists. Before adding further recurrence variables, functions should be refactored to pass the existing `struct rpt` instead. The structure can then be extended as needed.

**Comment by lfos** [comment]
> > If we want to implement the original request, we should first come up with a way to extend the data file format in a clean and backwards compatible way.
> 
> I have studied the recurrence rules in RFC 5545. In their entirety they are a daunting task, but parts of them would be a useful extention to calcurse. I'm thinking of (parts of) BYDAY, BYMONTHDAY and BYMONTH. The basic FREQ, INTERVAL and UNTIL rules are already implemented, and because the others are optional, an extension of the existing data file format that is backwards compatible, should pose no problem.

Sure. We need to make sure we don't sacrifice too much of the simplicity of the data file format, though. So how exactly do you suggest to extend the format?

**Comment by lhca** [comment]
> Sure. We need to make sure we don't sacrifice too much of the simplicity of the data file format, though. So how exactly do you suggest to extend the format?

The recurrence information for an event or appointment is enclosed in brackets like this:

`{1W -> 04/17/2019 !02/13/2019 !04/10/2019 !04/17/2019}`

First comes the required INTERVAL/FREQ (in calcurse terms frequency/type), then an optional until date flagged by `->`, then an optional list of one or more exception days each flagged by`!`.

Each of BYDAY, BYMONTHDAY and BYMONTH is an optional list of weekdays, monthdays and months respectively. The format of each list element need not concern us here, except that it's a short character string. The simplest is BYMONTH where the list elements are integers 1, 2, ..., 12.

I propose to follow the exception day pattern and describe each rule by an optional list of rule elements, each flagged by a single letter, `m` for BYMONTH, `d` for BYMONTHDAY and `w` for BYDAY. For example

`{1W -> 04/17/2019 !02/13/2019 !04/10/2019 !04/17/2019 m3 m10}`

would reduce the weekly occurrence to March and October. The two other rules are more complicated, but the format similar. For example

`{1W -> 04/17/2019 !02/13/2019 !04/10/2019 !04/17/2019 wSA wSU}`

would expand the single weekly occurrence to two (on Saturday and Sunday) and

`{2M m1 m15 m-1}`

would expand a single occurrence every other month to three occurrences on the first, fifteenth and last day of the month, every other month.

I do not know how much is doable or worth the effort, but that should not stop us from agreeing on a
format.

**Comment by lfos** [comment]
Sounds like a sensible first suggestion. I wonder whether we can make it slightly more consistent with our current implementation where the general description of the repetition type comes first, follwed by the end date, followed by exceptions. Can we put the new repetition rules next to the interval/frequency specifier?

Like `{1W wSA wSU -> 04/17/2019 !02/13/2019 !04/10/2019 !04/17/2019}`?

**Comment by lhca** [comment]
RFC5545 explicitly defines that the rule parts are not ordered in any particular sequence in the iCalendar data format.

But I so no reason why the calcurse file format should adopt that requirement. The optional parts may appear in any sequence as long as that sequence remains fixed. For example, first BYDAY, followed by BYMONTHDAY, BYMONTH, UNTIL and EXDATE.

**Comment by lfos** [comment]
Sounds good to me!

**Comment by bratekarate** [comment]
I am not sure if I understand the explanation at the end of #258 properly.

So the extensions for recurrence rules have not been added in 4.6, but are they planned for 4.7? Or does this mean development in this direction is postponed indefinitely? Sorry if this was answered in said PR, but I was honestly unable to extract that information.

I know this issue is not closed and therefore not rejected, just wondering if it is reasonable to expect it being added in some of the next releases or not.

**Comment by lfos** [comment]
> So the extensions for recurrence rules have not been added in 4.6, but are they planned for 4.7? Or does this mean development in this direction is postponed indefinitely? Sorry if this was answered in said PR, but I was honestly unable to extract that information.

The patches in that PR have been merged and will be part of the next release.

**Comment by bratekarate** [comment]
> The patches in that PR have been merged and will be part of the next release.

Oh, thank you for letting me now. Somehow I managed to read it as if everything but the support for the other recurrence types was merged  :facepalm:. Sorry about the confusion
